### DISCUSSION TITLE
How to render "nothing" based on a condition?

### DISCUSSION THREAD

**Discussion by amirhhashemi** [ueberdosis/tiptap#2852] [General Questions] [answered]
I'm extending the image extension to add support for uploading to the server, caption, etc and I'm doing this with NodeViews. When an image is added, I generate a data-url for the image and set it as the image `src` so the users can immediately see the selected image on the editor. Meanwhile, I send the image to the server and when it's uploaded I update the image `src` to a http url. I also save the editor content to a database as something is typed. 

The problem is that when the image is not fully uploaded to the server, its `src` is a giant data-url and I don't want to send that to the server. Basically, I don't want the image to be in the output when its `src` is a data-URL. So how can I do that?

I've tried this but I get wired errors:
```tsx
        renderHTML({ HTMLAttributes }) {
          if ((HTMLAttributes.src as string).startsWith("data:")) {
            return null; // I also tried returning undefined, [""] and []
          }

          // ...
        },
```

**Comment by pointnet** [comment]
How about replacing it by a 0x0 transparent png?

```javascript
renderHTML({ HTMLAttributes }) {
  if (HTMLAttributes.src.startsWith("data:")) {
    return [
      "img",
      mergeAttributes(this.options.HTMLAttributes, {
        src: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=',
        width: 0,
        height: 0
      }),
      0
    ];
  }

  // ...
},
```

**Reply by amirhhashemi** [reply]
Yeah, I think that works but I don't want images with data-url at all. That way I just remove an image with data-url and add another one. I would rather render an empty paragraph that contains the image caption or just the file name. Thanks BTW.

**Comment by amirhhashemi** [comment]
**Ignore this, just read the pointnet's reply below**
I did more research. We **must** return at least one HTML tag from `renderHTML` (`toDOM` in ProseMirror) so **the answer to my question is NO**. If you don't want images with data-url you can instead render an empty paragraph that contains the image caption or just the file name or a message that says the image is not uploaded successfully. Of course, I'm talking about the html output, you probably don't want such messages in the editor.

**Reply by pointnet** [reply]
**The answer to your question is YES!**

You have to "trick" `prosemirror-view` to let him believe your textnode is in fact an htmlement ;)

```javascript
  renderHTML({ HTMLAttributes }) {
    if (HTMLAttributes.src.startsWith("data")) {
      const dom = document.createTextNode("");
      dom.hasAttribute = () => true;
      return dom;
    }

    return ["img", HTMLAttributes];
  },
```

Complete sandbox: https://codesandbox.io/s/nostalgic-matan-xj876x

**Reply by amirhhashemi** [reply]
Yeah, it works! thank you again. You saved me 2 days so far:)

**Comment by pointnet** [comment, marked as answer]
You have to "trick" `prosemirror-view` to let him believe your textnode is in fact an htmlement ;)

```javascript
  renderHTML({ HTMLAttributes }) {
    if (HTMLAttributes.src.startsWith("data")) {
      const dom = document.createTextNode("");
      dom.hasAttribute = () => true;
      return dom;
    }

    return ["img", HTMLAttributes];
  },
```

Complete sandbox: https://codesandbox.io/s/nostalgic-matan-xj876x
